File: backend/schemas/analysis.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel


class PriceVsDistrictOut(BaseModel):
    id: str
    title: str
    district: Optional[str] = None
    price: Optional[float] = None
    district_median: Optional[float] = None
    difference: Optional[float] = None
    difference_percent: Optional[float] = None


class InitialCostOut(BaseModel):
    id: str
    title: str
    total_monthly_cost: Optional[float] = None
    deposit: Optional[float] = None
    initial_cost: Optional[float] = None
    initial_cost_is_estimate: bool


class DistrictProfileOut(BaseModel):
    district: str
    count: int
    median_price: Optional[float] = None
    median_total_monthly_cost: Optional[float] = None
    median_deposit: Optional[float] = None
    negotiable_percent: Optional[float] = None
    no_deposit_percent: Optional[float] = None
    has_additional_cost_percent: Optional[float] = None


class OutlierBoundsOut(BaseModel):
    q1: Optional[float] = None
    q3: Optional[float] = None
    iqr: Optional[float] = None
    lower_bound: Optional[float] = None
    upper_bound: Optional[float] = None


class OutlierItemOut(BaseModel):
    id: str
    title: str
    metric: str
    value: float


class OutliersOut(BaseModel):
    metric: str
    bounds: Optional[OutlierBoundsOut] = None
    outliers: list[OutlierItemOut]


class ValueScoreComponentsOut(BaseModel):
    price_z_score: Optional[float] = None
    total_cost_z_score: Optional[float] = None
    negotiable_bonus: float
    deposit_penalty: float


class ValueScoreOut(BaseModel):
    id: str
    title: str
    district: Optional[str] = None
    value_score: Optional[float] = None
    reason: Optional[str] = None
    components: Optional[ValueScoreComponentsOut] = None

File: frontend/src/types/analysis.ts
/** Types mirroring `backend/schemas/analysis.py`. */

export type AnalysisMetric = "price" | "total_monthly_cost" | "additional_cost" | "deposit";

export interface PriceVsDistrict {
  id: string;
  title: string;
  district: string | null;
  price: number | null;
  district_median: number | null;
  difference: number | null;
  difference_percent: number | null;
}

export interface InitialCost {
  id: string;
  title: string;
  total_monthly_cost: number | null;
  deposit: number | null;
  initial_cost: number | null;
  initial_cost_is_estimate: boolean;
}

export interface DistrictProfile {
  district: string;
  count: number;
  median_price: number | null;
  median_total_monthly_cost: number | null;
  median_deposit: number | null;
  negotiable_percent: number | null;
  no_deposit_percent: number | null;
  has_additional_cost_percent: number | null;
}

export interface OutlierBounds {
  q1: number | null;
  q3: number | null;
  iqr: number | null;
  lower_bound: number | null;
  upper_bound: number | null;
}

export interface OutlierItem {
  id: string;
  title: string;
  metric: string;
  value: number;
}

export interface Outliers {
  metric: string;
  bounds: OutlierBounds | null;
  outliers: OutlierItem[];
}

export interface CostDistributionBucket {
  from: number;
  to: number;
  count: number;
}

export interface ValueScoreComponents {
  price_z_score: number | null;
  total_cost_z_score: number | null;
  negotiable_bonus: number;
  deposit_penalty: number;
}

export interface ValueScore {
  id: string;
  title: string;
  district: string | null;
  value_score: number | null;
  reason: string | null;
  components: ValueScoreComponents | null;
}
